dtm0/service.c

```diff
@@ -0,0 +1,648 @@
+						    m0_get()->i_dtm0_log_key);
+		M0_ASSERT(dtm0->dos_log != NULL);
+	}
+
+out:
+	return m0_be_dtm0_log_init(&dtm0->dos_log, &dtm0->dos_clk_src,
+				   dtm0->dos_origin == DTM0_ON_PERSISTENT);
+}
+
+static int dtm0_service_start(struct m0_reqh_service *service)
+{
+        M0_PRE(service != NULL);
+        return dtm_service__origin_fill(service) ?:
+		dtm0_service__ha_subscribe(service);
+}
+
+static void dtm0_service_stop(struct m0_reqh_service *service)
+{
+	struct m0_dtm0_service *dtm0;
+
+	M0_PRE(service != NULL);
+	dtm0 = to_dtm(service);
+
+	dtm0_service__ha_unsubscribe(service);
+
+	m0_dtm0_fop_fini();
+	/**
+	 * It is safe to remove any remaining entries from the log
+	 * when a process with volatile log is going to die.
+	 */
+	if (dtm0->dos_origin == DTM0_ON_VOLATILE && dtm0->dos_log != NULL) {
+		m0_be_dtm0_log_clear(dtm0->dos_log);
+		m0_be_dtm0_log_fini(&dtm0->dos_log);
+	}
+}
+
+static void dtm0_service_fini(struct m0_reqh_service *service)
+{
+	M0_PRE(service != NULL);
+	dtm0_service__fini(to_dtm(service));
+        m0_free(service);
+}
+
+M0_INTERNAL int m0_dtm0_stype_init(void)
+{
+	extern struct m0_sm_conf m0_dtx_sm_conf;
+
+	return m0_sm_addb2_init(&m0_dtx_sm_conf,
+				M0_AVI_DTX0_SM_STATE, M0_AVI_DTX0_SM_COUNTER) ?:
+            m0_dtm0_fop_init() ?:
+		    m0_reqh_service_type_register(&dtm0_service_type);
+}
+
+M0_INTERNAL void m0_dtm0_stype_fini(void)
+{
+	m0_reqh_service_type_unregister(&dtm0_service_type);
+}
+
+M0_INTERNAL bool m0_dtm0_is_a_volatile_dtm(struct m0_reqh_service *service)
+{
+	return m0_streq(service->rs_type->rst_name, "M0_CST_DTM0") &&
+		to_dtm(service)->dos_origin == DTM0_ON_VOLATILE;
+}
+
+M0_INTERNAL bool m0_dtm0_is_a_persistent_dtm(struct m0_reqh_service *service)
+{
+	return m0_streq(service->rs_type->rst_name, "M0_CST_DTM0") &&
+		to_dtm(service)->dos_origin == DTM0_ON_PERSISTENT;
+}
+
+M0_INTERNAL struct m0_dtm0_service *
+m0_dtm0_service_find(const struct m0_reqh *reqh)
+{
+	struct m0_reqh_service *rh_srv;
+
+	rh_srv = m0_reqh_service_find(&dtm0_service_type, reqh);
+
+	return rh_srv == NULL ? NULL : to_dtm(rh_srv);
+}
+
+M0_INTERNAL bool m0_dtm0_in_ut(void)
+{
+	return M0_FI_ENABLED("ut");
+}
+
+/* --------------------------------- EVENTS --------------------------------- */
+
+static int ready_to_process_ha_msgs = 0;
+
+static bool service_connect_clink(struct m0_clink *link)
+{
+	M0_ENTRY();
+	M0_LOG(M0_DEBUG, "DTM0 service: connected");
+	M0_LEAVE();
+	return true;
+}
+
+static void service_connect_ast(struct m0_sm_group *grp,
+				struct m0_sm_ast   *ast)
+{
+	struct dtm0_process *process = ast->sa_datum;
+	int rc;
+
+	M0_ENTRY();
+	m0_clink_init(&process->dop_service_connect_clink, service_connect_clink);
+	process->dop_service_connect_clink.cl_is_oneshot = true;
+	rc = m0_dtm0_service_process_connect(process->dop_dtm0_service,
+					     &process->dop_rserv_fid,
+					     process->dop_rep,
+					     true);
+	M0_ASSERT(rc == 0);
```

Comment:
error handling

Reply:
we will not fix this in given patch, as new failure handing schema for connection manager will be introduced during start/stop milestone (next one).